File: 9.Minimum-Spanning-Tree_Shortest_Path/jiyeon/2.cheapest-flights-within-k-stops.py
# https://leetcode.com/problems/cheapest-flights-within-k-stops/

import collections
import heapq
from typing import List


class Solution:
    def findCheapestPrice(self, n, flights, start_point, end_point, stops):
        dicts = collections.defaultdict(dict)
        for curr_node, next_node, price in flights:
            dicts[curr_node][next_node] = price

        # Min-heap to store minimum weight edge at top.
        heap = [(0, start_point, stops + 1)]

        # Track nodes which are included in MST.
        visited = [0] * n

        while heap:
            price, idx, stops = heapq.heappop(heap)
            if idx == end_point:
                return price
            if visited[idx] >= stops:
                continue
            visited[idx] = stops
            for curr_stops, dict_price in dicts[idx].items():
                heapq.heappush(
                    heap, (price + dict_price, curr_stops, stops - 1))
        return -1


print(Solution().findCheapestPrice(4,
                                   [[0, 1, 100], [1, 2, 100], [2, 0, 100],
                                    [1, 3, 600], [2, 3, 200]],
                                   0,
                                   3,
                                   1))

File: 9.Minimum-Spanning-Tree_Shortest_Path/동규/2.cheapest-flights-within-k-stops.js
// https://leetcode.com/problems/cheapest-flights-within-k-stops/

var findCheapestPrice = function (n, flights, src, dst, k) {
  let prevArr = new Array(n).fill(Infinity);
  let currentArr = new Array(n).fill(Infinity);
  prevArr[src] = 0;
  currentArr[src] = 0;
  for (let i = 0; i < k + 1; i++) {
    for (let [from, to, price] of flights) {
      if (
        price + prevArr[from] < prevArr[to] &&
        price + prevArr[from] < currentArr[to]
      ) {
        currentArr[to] = price + prevArr[from];
      }
    }
    prevArr = [...currentArr];
  }
  return currentArr[dst] === Infinity ? -1 : currentArr[dst];
};
